### PR TITLE
[google-cloud-cpp]: support for v2.x.y series

### PR DESCRIPTION
**Pull request by coryan**
This change splits packaging for the v2.x vs. v1.x series.  The 2.x series requires C++14, and include dozens of new components. I have updated the package to use the `googleapis` package in Conan.

Specify library name and version:  **google-cloud-cpp/2.5.x**

The current recipe does not work. It tries to "unvendor" googleapis, but did not complete the work.  It also does not compile all the components.  I am one of the maintainers for `google-cloud-cpp` upstream.  FWIW, I think the next release will have fewer (or at least smaller) patches.

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by coryan** [comment]
The errors seem to be caused by low disk space on the build runners.  Can somebody take a look?

**Comment by prince-chrismc** [comment]
Poking v2 pipeline

File: recipes/google-cloud-cpp/2.x/conanfile.py
Comment (prince-chrismc) [inline recipes/google-cloud-cpp/2.x/conanfile.py:89]:
```suggestion
        self.requires("protobuf/3.21.9")
```

File: recipes/google-cloud-cpp/2.x/conanfile.py
Comment (coryan) [reply recipes/google-cloud-cpp/2.x/conanfile.py]: Thanks for these changes.  Just for my education: doesn't this need to wait until gRPC gets the same bump in #16564 ?

**Comment by prince-chrismc** [comment]
needs #16185 blah

**Comment by coryan** [comment]
Is there anything I could do to speed this along?

**Comment by jcar87** [comment]
Hi @coryan - thanks for checking in with us.

Now that https://github.com/conan-io/conan-center-index/pull/16185 has been merged, this is also in the process of being built: https://github.com/conan-io/conan-center-index/pull/16564

and then we can re-run this one!

**Comment by jcar87** [comment]
I believe the dependency conflicts have now been resolved - will check CI results once they're published

**Comment by coryan** [comment]
Is the conan v2 error caused because grpc is not in [this list](https://github.com/conan-io/conan-center-index/blob/master/.c3i/conan_v2_ready_references.yml)?

**Comment by prince-chrismc** [comment]
Still does not work in v2 https://github.com/conan-io/conan-center-index/pull/16564#issuecomment-1505921507 :)

File: recipes/google-cloud-cpp/2.x/conanfile.py
Comment (prince-chrismc) [inline recipes/google-cloud-cpp/2.x/conanfile.py:30]:
```suggestion
    package_type = "library"
    settings = "os", "arch", "compiler", "build_type"
```

File: recipes/google-cloud-cpp/2.x/conanfile.py
Comment (prince-chrismc) [inline recipes/google-cloud-cpp/2.x/conanfile.py:33]: In dangerous to have this and call it in the generator method since they conflict and overwrite 

```suggestion
```

File: recipes/google-cloud-cpp/2.x/conanfile.py
Comment (prince-chrismc) [inline recipes/google-cloud-cpp/2.x/conanfile.py:51]: If these are required then it should be checked in the validate method as well

File: recipes/google-cloud-cpp/2.x/conanfile.py
Comment (prince-chrismc) [inline recipes/google-cloud-cpp/2.x/conanfile.py:122]: This is deprecated and does not need to be supported, we can remove it if you dont need it :)

**Review by prince-chrismc — COMMENTED** [review]
Couple of small nits but I think with this much work its better to get it in and polish it.

Build is passing so should be more helpful then not having it at all

File: recipes/google-cloud-cpp/2.x/conanfile.py
Comment (coryan) [reply recipes/google-cloud-cpp/2.x/conanfile.py:31]: Fixed

File: recipes/google-cloud-cpp/2.x/conanfile.py
Comment (coryan) [reply recipes/google-cloud-cpp/2.x/conanfile.py]: Removed

File: recipes/google-cloud-cpp/2.x/conanfile.py
Comment (coryan) [reply recipes/google-cloud-cpp/2.x/conanfile.py:51]: Fixed

File: recipes/google-cloud-cpp/2.x/conanfile.py
Comment (coryan) [reply recipes/google-cloud-cpp/2.x/conanfile.py]: Candidly I am not sure if I need it.  Local testing with v2 seems to require a `VirtualRunEnv()` call, but I may be doing something else wrong.  I would appreciate any guidance here.

**Comment by coryan** [comment]
PTAL.

**Comment by RubenRBS** [comment]
Hi @coryan I've generated the missing packages that were making the v2 pipeline fail. This has now shown a conflict between your recipe and libcrul's. I'll try to solve this for you tomorrow so that we can merge your recipe. Thanks a lot for your patience :)

**Comment by coryan** [comment]
@RubenRBS thanks for being proactive. I am really looking forward to getting this merged.

**Comment by RubenRBS** [comment]
Update: We're getting ready to do some repo-wide work regarding Openssl dependencies which will also solve the conflict issues you're seeing, but it might take a few days until we process everything that's needed for it. I'll keep you posted on the status once that's done so we can move the recipe forward :)

**Comment by coryan** [comment]
@RubenRBS thanks for the update.

Maybe you can help me calibrate my expectations for future PRs:  I started the work preparing for this PR almost 4 months ago (see #14894).  Maybe I just go unlucky, with the holiday break, and the upgrades to conan v2.  But it feels that I should plan that updates to `google-cloud-cpp` will take multiple weeks, as they require (at a minimum) PRs to `googleapis`, `grpc-proto`, and `grpc` before sending an update to `google-cloud-cpp`.  And that is if I am not unlucky and other deps (e.g. `protobuf`, `libcurl`, or `grpc` itself) change in the interim.

I am thinking that I should send PRs to update `google-cloud-cpp` every 3 months or so, trying to go faster than that is just not going to be a good use of anybody's time.  Does that sound reasonable?

Thanks in advance,

**Comment by RubenRBS** [comment]
Hi @coryan thanks for your question, I've asked @jcar87 too provide a better answer that I ever could :)

**Comment by jcar87** [comment]
Hi @coryan - apologies for the time it has taken to review and merge the PRs related to google-cloud-cpp and related libraries. Indeed we have been focusing on Conan 2.0 compatibility of recipes - ensuring that more and more recipes work and have packages published for Conan 2.0 users.  

Now that Conan 2.0 is out of the door and more and more recipes are compatible (we estimate around 1000 are already compatible, with a little over 400 having published packages), we are focusing on going through the PR backlog. Apologies this is taking longer than we would like.

Thank you for your efforts maintaining this recipe - in the past there have been issues related to both the vendoring-in of compiled protos, as well as changes to the version of Protobuf (or other dependencies). We are currently contemplating solutions to avoid a cascade of PR fixes to align versions and start using version ranges, as well as re-create any packages - although I suspect that aligning protobuf, grpc, grpc-proto, google-apis and google-cloud-cpp may not be trivial. I'd be very interested in learning about the release cadence and compatibility constraints so that we can take the necessary measures to ensure consumers of these libraries have a smooth experience, as well as making the job of contributors and maintainers as easy as possible. 

In the meantime we are currently looking into ensuring the packages for the correct combination of dependencies are ready to unblock this PR.

**Comment by coryan** [comment]
Thank you.  Seems like things will get better in the future.  I will keep working on this PR and start a new series as soon as this is done.

> I'd be very interested in learning about the release cadence and compatibility constraints so that we can take the necessary measures to ensure consumers of these libraries have a smooth experience, as well as making the job of contributors and maintainers as easy as possible.

I will try to summarize (I am usually quite bad). As to release cadence

* `Protobuf` does not currently have a predictable release cadence, but they are working towards one.  I spect it will be in the "once every few months".
* `googleapis` does not have a release cadence.  It does not have releases.  Generally you can expect multiple changes per week.  Candidly, nobody at Google thinks of it as a "package".  Is just a list of specifications for public APIs, more on that below.
* `grpc-proto` does not have release cadence, nor does it have many changes.
* `grpc` releases every 6 weeks.
* `google-cloud-cpp` releases once a month.

As to the compatibility constraints:

* `grpc-proto` and `grpc` can use any version of `googleapis`.  The dependencies on it are relatively small and very stable.
* `google-cloud-cpp` must be compiled with the version of `googleapis` that it "vendors".  There is code in `google-cloud-cpp` that will assume classes generated from the `.proto` files have specific functions.  It may be possible to use newer versions of `googleapis`, but even that is risky: the `googleapis` do remove functions from time to time.

In general, `googleapis` is not thought of as a "library" or "package" in Google. It is just a list of the (current) network-level API specifications.

**Comment by jcar87** [comment]
hi @coryan - thank you so much for the clarifications.

If memory serves me well, `googleapis` was split to a separate package (to be consumed by grpc and `google-cloud-cpp`, e.g. https://github.com/conan-io/conan-center-index/pull/12062) to minimise the possibility of having symbols compiled from `pb.cc` files generated from same `.proto` files (from googleapis) ending up in multiple libraries - to avoid any issues in case `google-cloud-cpp` and `grpc` `and `grpc-proto` all used different versions of the protos from googleapis. I believe there was at least one case where this was suspected to have happened whe

Do you know if this is a possibility? We are not particularly committed to the idea of a standalone `googleapis` package if there is an easier way to solve this or avoid these sort of compatibility issues. My understanding of the conventions of Protobuf specifications it that messages are kept wire/backwards compatible, but here we have an extra dimension that may have to do more with symbol visibility and ODR violations than with a protobuf message having a breaking change.

**Comment by coryan** [comment]
> If memory serves me well, `googleapis` was split to a separate package (to be consumed by grpc and `google-cloud-cpp`, e.g. #12062) to minimise the possibility of having symbols compiled from `pb.cc` files generated from same `.proto` files (from googleapis) ending up in multiple libraries - to avoid any issues in case `google-cloud-cpp` and `grpc` `and `grpc-proto` all used different versions of the protos from googleapis. I believe there was at least one case where this was suspected to have happened whe


> Do you know if this is a possibility? 

Not really.

- Most of the proto files that are used in gRPC public libraries are not compiled with `protoc` to generate `.pb.h` files. They are compiled with [`upb`](https://github.com/protocolbuffers/upb) and generate `.upb.h` files.  These files have different symbols and will not conflict with any of the symbols in `google-cloud-cpp`.
   - But note that grpc vendors-in both the upb-generated files and upb itself.
   - That will be a problem if Protobuf takes a dependency on `upb` or if `upb` is released as a separate package.
   - I have no secret knowledge as to whether either of these things is possible.
- Only a few proto files are compiled into `.pb.h` files and used in public libraries from `grpc`.  Most are used only in tests.
- The two libraries that do this are `grpcpp_channelz` and `grpc++_reflection` libraries.
    - They are  public and do have some `*.pb.h` (and `*.grpc.pb.h` files).
    - None of these files are used in `google-cloud-cpp`.

I cannot see a way to produce a ODR-violation by linking both `google-cloud-cpp` and `grpc`. 

>We are not particularly committed to the idea of a standalone `googleapis` package if there is an easier way to solve this or avoid these sort of compatibility issues.

Interesting.

> My understanding of the conventions of Protobuf specifications it that messages are kept wire/backwards compatible, but here we have an extra dimension that may have to do more with symbol visibility and ODR violations than with a protobuf message having a breaking change.

I appreciate that.  I think this is less of a problem than usual.  Basically gRPC goes out of its way to avoid exporting symbols that may conflict with `google-cloud-cpp` (or any C++ application using protos from `googleapis`).

**Comment by jcar87** [comment]
Hi @coryan - thanks for your reply, these are very valuable insights! Based on this I have further questions.

From what I can see, the gPRC CMake scripts download googleapis, but I can't really see any of the .proto files inside being generated at all - and the `upb`-generated files appear to be already bundled with grpc itself - is this correct? I wonder what the googleapis contents are needed for, if at all, in order to build grpc?

I also can't find obvious references to protos from googleapis in `grpc-proto` - would seem the ones required are the ones from protobuf itself?

File: recipes/google-cloud-cpp/2.x/conanfile.py
Comment (jcar87) [inline recipes/google-cloud-cpp/2.x/conanfile.py:106]:
```suggestion
        self.requires("openssl/[>=1.1 <4]")
```

We've updated the `grpc` recipe to use a version range instead, and fixing it here to a version range should hopefully resolve version conflicts (unless anywhere else in the graph is trying to use a specific version, which we can also address).

**Comment by coryan** [comment]
Woo hoo!  All the tests pass.  It seems this is ready?

**Review by SSE4 — APPROVED** [review]

**Comment by jjcasmar** [comment]
Is this ready to go?

**Comment by coryan** [comment]
> Is this ready to go?

I think so.  Unless @jcar87 or @RubenRBS suggest some changes to this PR.

**Comment by coryan** [comment]
Ping?

**Review by jcar87 — APPROVED** [review]
Thanks so much!! LGTM

**Pull request merged**
